### ISSUE TITLE
CKV_AWS_111: IAM fails even when a condition exists

### ISSUE THREAD

**Issue by davops** [bridgecrewio/checkov#5305]
**Describe the issue**
An IAM policy check is failing even when there is a constraint condition.
Error:
```
Check: CKV_AWS_111: "Ensure IAM policies does not allow write access without constraints"
FAILED for resource: aws_iam_policy_document.myKmsKey
```

**Examples**
Terraform:
```
data "aws_iam_policy_document" "myKmsKey" {
    statement {
    actions = [
      "kms:Decrypt"
    ]
    resources = [
      "*"
    ]

    condition {
      test     = "ArnEquals"
      variable = "aws:SourceArn"
      values = [
        "foo"
      ]
    }
  }
}
```

**Version (please complete the following information):**
 - Checkov Version 2.3.314

**Additional context**
It is the same as this issue: https://github.com/bridgecrewio/checkov/issues/5148 but does not appear to be resolved.

**Comment by gruebel** [comment]
hey @davops thanks for reaching out.

They are quite similar, but test for a bit different things. `CKV_AWS_111` looks for an IAM policy which allows a writing operations to all resources, conditions don't matter, because it still gives permission to do the action against all resources.

We don't check to what you attach the data block and therefore it can result in a false positive, if it is attached to a resource instead of an identity.

**Comment by davops** [comment]
Thank you @gruebel for the information. It's still giving us a false positive in certain scenarios. For instance, according to AWS's documentation[ here](https://docs.aws.amazon.com/kms/latest/developerguide/alias-authorization.html), the resource should be a wildcard when using a KMS Request Alias on a policy attached to a role. Specifically, it says "These features do not allow you to identify a KMS key by using an alias in the resource element of a policy statement." 
Thus, this Terraform example would fail that check:
```
data "aws_iam_policy_document" "test" {
  statement {
    sid = "Key policy using a request alias condition"

    actions = [
      "kms:Decrypt"
    ]

    resources = ["*"]

    condition {
      test     = "StringEquals"
      variable = "kms:RequestAlias"
      values = [
        "alias/*alpha*",
      ]
    }
  }
}
```
How can we avoid a false positive in this case?

**Comment by gruebel** [comment]
just reduce the resource scope to all keys in the account 🙂 
```hcl
data "aws_iam_policy_document" "test" {
  statement {
    sid = "Key policy using a request alias condition"

    actions = [
      "kms:Decrypt"
    ]

    resources = ["arn:aws:kms:us-west-2:111122223333:key/*"]

    condition {
      test     = "StringEquals"
      variable = "kms:RequestAlias"
      values = [
        "alias/*alpha*",
      ]
    }
  }
}
```

**Comment by bdoyle0182** [comment]
@gruebel that solution doesn't work for all cases and multi-account setups where the condition is something like an enterprise's org id and you don't have any other option but to wildcard account numbers. Is there anything else we can do to improve this check? It's our largest false positive check

see another discussion on this topic. it seems like a change was made to cloudsplaining that should make * plus condition keys being set makes a passing check.
https://gitlab.com/bridgecrewio/checkov/-/issues/990
https://github.com/salesforce/cloudsplaining/pull/167

but not seeing this behavior in checkov

edit: this is specifically for cloudformation. it seems like for terraform there was a bugfix in 2023 to pass checks for resource * if condition fields are set
https://github.com/bridgecrewio/checkov/pull/5171/files
